gnomad_qc/v4/sample_qc/relatedness.py

```diff
@@ -0,0 +1,251 @@
+    test = args.test
+
+    if args.print_cuking_command:
+        print(
+            textwrap.dedent(
+                f"""\
+                # Make sure that $PROJECT_ID is set.
+                cd cuKING && \\
+                ./cloud_batch_submit.py \\
+                    --location=us-central1 \\
+                    --project_id=$PROJECT_ID \\
+                    --tag_name=$(git describe --tags) \\
+                    --input_uri={cuking_input_path(test=test)} \\
+                    --output_uri={cuking_output_path(test=test)} \\
+                    --requester_pays_project=$PROJECT_ID \\
+                    --kin_threshold={args.second_degree_kin_cutoff} \\
+                    --split_factor=4 && \\
+                cd ..
+                """
+            )
+        )
+        return
+
+    hl.init(
+        log="/relatedness.log",
+        default_reference="GRCh38",
+    )
+
+    try:
+        if args.prepare_inputs:
+            input_path = cuking_input_path(test=test)
+
+            if hadoop_exists(input_path):
+                raise DataException(f"{input_path} already exists")
+
+            mt_v3 = hl.read_matrix_table(
+                get_predetermined_qc(version="3.1", test=test).path
+            )
+            mt_v4 = hl.read_matrix_table(get_predetermined_qc(test=test).path)
+
+            # Check that there are no duplicate sample names.
+            sample_list = mt_v3.s.collect() + mt_v4.s.collect()
+            if len(sample_list) != len(set(sample_list)):
+                raise DataException("duplicate samples found in v3 + v4")
+
+            # Check that the number of sites match.
+            row_count_v3 = mt_v3.count_rows()
+            row_count_v4 = mt_v4.count_rows()
+            if row_count_v3 != row_count_v4:
+                raise DataException("number of sites doesn't match between v3 and v4")
+
+            # Combine v3 and v4 dense QC matrices, as we need to compute relatedness
+            # across the union of samples.
+            mt = mt_v3.union_cols(mt_v4)
+
+            # Compute the field required for KING.
+            mt = mt.select_entries(n_alt_alleles=mt.GT.n_alt_alleles())
+
+            # Create a table based on entries alone. By adding a row and column index,
+            # we can drop the locus, alleles, and sample field, as well as skip writing
+            # missing entries.
+            mt = mt.select_globals()
+            mt = mt.add_row_index()
+            mt = mt.add_col_index()
+            entries = mt.entries()
+            entries = entries.key_by()
+            entries = entries.select(
+                entries.row_idx, entries.col_idx, entries.n_alt_alleles
+            )
+
+            # Export one compressed Parquet file per partition.
+            logger.info("Writing Parquet tables...")
+            entries.to_spark().write.option("compression", "zstd").parquet(input_path)
+
+            # Write metadata that's useful for postprocessing. Map `col_idx` to `s`
```

Comment:
cuKING uses sample indices instead of sample strings in the input data, to reduce size. The output contains sample strings though, for making the import easier. I'll improve the comment in the code if we decide to go with calling `dense_mt_to_parquet.py` directly.